File: jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/accmod/PrefixedPropertiesProvider.java
package org.jenkinsci.gradle.plugins.jpi2.accmod;

import org.gradle.api.Project;
import org.gradle.api.provider.Provider;

import java.util.Map;
import java.util.stream.Collectors;

/**
 * Extracts Gradle project properties that share a common prefix and strips that prefix,
 * making them available as a plain map to the access-modifier checker.
 */
public class PrefixedPropertiesProvider {
    private PrefixedPropertiesProvider() {
    }

    /**
     * @param project the Gradle project whose properties are scanned
     * @param prefix  only properties with this prefix are included; the prefix is stripped from the keys
     * @return a provider with a map of properties with the given prefix, with the prefix stripped from the keys
     */
    public static Provider<Map<String, String>> gradlePropertiesPrefixedBy(Project project, String prefix) {
        return project.getProviders().gradlePropertiesPrefixedBy(prefix).map(properties ->
                properties.entrySet().stream()
                        .filter(e -> e.getValue() != null)
                        .collect(Collectors.toMap(
                                e -> e.getKey().substring(prefix.length()),
                                Map.Entry::getValue))
        );
    }
}

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/PrefixedPropertiesProvider.kt
package org.jenkinsci.gradle.plugins.accmod

import org.gradle.api.Project
import java.util.concurrent.Callable

open class PrefixedPropertiesProvider(private val project: Project, private val prefix: String) : Callable<Map<String, Any>> {
    override fun call(): Map<String, Any> {
        val filtered = mutableMapOf<String, Any>()
        for ((key, value) in project.properties) {
            if (key.startsWith(prefix) && value != null) {
                val tidy = key.replace(prefix, "")
                filtered[tidy] = value
            }
        }
        return filtered
    }
}
